### DISCUSSION TITLE
Le attività eliminate ricompaiono (cap 8 e 9)

### DISCUSSION THREAD

**Discussion by Claudia-Spina** [lifeisfoo/react-guida-pratica#18] [Errori e refusi]
Nell'applicazione sviluppata nel volume, se si elimina un'attività e poi si modifica lo stato di un'altra attività tramite la checkbox, l'attività eliminata ricompare in elenco. Come mai? Tra l'altro, se l'attività eliminata era non completata, la sua nuova visualizzazione, unita al numero segnalato dal relativo contatore delle attività non completate (numero che nella situazione descritta può addirittura diventare negativo), può ingenerare confusione. Comunque ottimo libro! Sono alle prime armi nel mondo della programmazione e dello sviluppo front-end, e mi sono trovata benissimo.

**Comment by lifeisfoo** [comment]
Ciao Claudia,
sono contento che il libro abbia soddisfato le tue aspettative, soprattuto visto che sei alle prime armi.

Riguardo il problema che descrivi:
- ti riferisci al codice dell'ultimo capitolo o a quello di uno di quelli precedenti?
- riesci a condividere qui un file zip con tutta la cartella (escludendo i node_modules) del codice che stai usando?

**Reply by Claudia-Spina** [reply]
Domani controllerò meglio ed eventualmente invierò il file, ma mi sembra che il problema riguardi sia il codice dell'ultimo capitolo, sia quello dei capitoli precedenti (sin dall'introduzione della funzionalità di eliminazione delle attività), e se non erro era già presente prima che iniziassi a modificare il codice. Chiedo scusa in anticipo se sto riportando inesattezze.

**Reply by Claudia-Spina** [reply]
Ho provato a vedere se il problema si verificasse anche usando il codice relativo a CRA, posizionandomi in una cartella a caso (ch09) tra quelle in cui è implementata la funzione di eliminazione, e lo riscontro anche lì. Per lo sviluppo dell'applicazione ho utilizzato Vite, quindi non credo di aver modificato il codice nelle cartelle del branch di default "master". Dunque il codice a cui mi riferisco, salvo mie sviste, è quello del libro. Se serve, le versioni di Node che ho utilizzato sono state 10.16.3 con CRA e 18.10.0 con Vite.

**Reply by lifeisfoo** [reply]
Avevi ragione @Claudia-Spina: il bug c'è e sei la prima ad averlo individuato.

Nel capitolo 8 e 9 vengono usati due stati per tenere traccia di tutte le attività (`allTodos`) e di quelle legata alla singola lista selezionata (`todos`).

```
  const [allTodos, setAllTodos] = useState(initialTodos);
  const [todos, setTodos] = useState([]);
```

La funzione `handleDeleteTodo` elimina l'attività dall'array `todos` ma non da `allTodos` (ed è questo il problema).
La funzione `handleUpdateTodo` invece aggiorna l'attività in tutti e due gli array (corretto).
Quindi, quando elimini un'attività e poi cambi lo stato di un'altra, la funzione `handleUpdateTodo` aggiorna l'elenco delle attività da mostrare prendendole da `allTodos` dove è ancora presente quella che avevi eliminato.

Ho fatto un commit sia [sul branch di CRA](https://github.com/lifeisfoo/react-guida-pratica/commit/3cf97b3c68daf5d784cda5f3c77550619d27399f) che su [quello di Vite](https://github.com/lifeisfoo/react-guida-pratica/commit/63bfc053b6235eebb8086f1d5552658033dde73b) con la correzione.

Nel capitolo 10 e 11 il problema non dovrebbe essere presente perché lo stato di tutte le attività viene mantenuto nel server e non ci sono stati duplicati.

Purtroppo quando si scrivi del codice in modo incrementale per spiegare le cose in modo semplice è facile finire in situazioni come queste. Se riesco farò un aggiornamento con un'ulteriore versione semplificata della logica e un post per comunicarlo. Quindi tieni d'occhio sia questo repository che il sito.

Grazie di nuovo per aver segnalato il problema e fammi sapere se ora funziona in modo corretto.

**Reply by Claudia-Spina** [reply]
Sì, tutto giusto. Con questo accorgimento funziona bene. Grazie mille! Spiegazione chiarissima, come, del resto, tutto il libro.
